File: app/src/main/java/com/example/swahiliapplication/ui/SignInActivity.java
package com.example.swahiliapplication.ui;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatButton;

import com.example.swahiliapplication.ConstantValues;
import com.example.swahiliapplication.R;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.ActionCodeSettings;

import java.util.Objects;
import java.util.Random;

import butterknife.BindView;

public class SignInActivity extends AppCompatActivity {
    private ConstantValues constantValues=new ConstantValues();

    EditText emailText;

    private String emailGet;
    AppCompatButton loginButton;
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sign_in);
        emailText=findViewById(R.id.email_id_text);
        loginButton=findViewById(R.id.login_btn);

    }



    @Override
    protected void onStart() {
        super.onStart();

        loginButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                    signInWithLink();
            }
        });
    }
    private void signInWithLink(){
        emailGet= Objects.requireNonNull(emailText.getText()).toString().trim();
        if(emailGet.isEmpty()){
            emailText.setError("Sorry. This is required");
            emailText.requestFocus();
            return;
        }
        ActionCodeSettings actionCodeSettings =
                ActionCodeSettings.newBuilder()
                        .setUrl("https://swahiliapplication-e6b01.firebaseapp.com/verify?uid=".concat(String.valueOf(new Random().nextInt(1000))))
                        // This must be true
                        .setHandleCodeInApp(true)
                        .setAndroidPackageName(
                                "com.example.swahiliapplication",
                                false, /* installIfNotAvailable */
                                null    /* minimumVersion */)
                        .build();
        ConstantValues.getFirebaseAuth().sendSignInLinkToEmail(emailGet,actionCodeSettings).addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if(task.isSuccessful()){
                    emailText.setText("", TextView.BufferType.EDITABLE);
                    Toast.makeText(SignInActivity.this, "Kindly check your mailbox for sign in instructions", Toast.LENGTH_SHORT).show();
                }else{
                    Toast.makeText(SignInActivity.this, task.getException().toString(),Toast.LENGTH_SHORT).show();
                }
            }
        });
        SharedPreferences sharedPreferences = getSharedPreferences("emailAddress", MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString("emailAddress", emailGet);
        editor.putString("extraAction", getIntent().getStringExtra("extraAction"));
        editor.apply();

    }
}

File: app/src/main/java/com/example/swahiliapplication/ui/SignInActivity.kt
package com.example.swahiliapplication.ui

import androidx.appcompat.app.AppCompatActivity
import android.widget.EditText
import androidx.appcompat.widget.AppCompatButton
import android.widget.ProgressBar
import android.widget.TextView
import com.google.firebase.auth.FirebaseAuth
import android.app.ProgressDialog
import android.content.Context
import android.os.Bundle
import com.example.swahiliapplication.R
import android.content.Intent
import android.text.TextUtils
import android.util.AttributeSet
import android.util.Patterns
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.example.swahiliapplication.ui.CreateAccountActivity
import com.google.android.gms.tasks.OnSuccessListener
import com.google.firebase.auth.AuthResult
import com.example.swahiliapplication.SwahiliLevels
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.tasks.OnCompleteListener

class SignInActivity : AppCompatActivity() {
    lateinit var emailEditText: EditText
    private lateinit var passwordEditText: EditText
    lateinit var loginBtn: AppCompatButton
    lateinit var createAccountBtnTextView: TextView
    lateinit var firebaseAuth: FirebaseAuth
    lateinit var mGoogleApiClient: GoogleApiClient

     fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.activity_sign_in,container,false)
        view.apply {
            loginBtn = findViewById(R.id.login_btn)
            emailEditText = findViewById(R.id.email_edit_text)
            passwordEditText = findViewById(R.id.password_edit_text)


            firebaseAuth = FirebaseAuth.getInstance()

            loginBtn.setOnClickListener { loginFun() }
        }
        return view
    }

    private fun loginFun() {
        val email = emailEditText.text.toString().trim()
        val password = passwordEditText.text.toString().trim()
        var error = 0



        if ( TextUtils.isEmpty(email) || TextUtils.isEmpty(password)) {
            error = 1
        }

        if(error!=1){

        }


        when (error) {
            1 -> {
                if(TextUtils.isEmpty(email))
                    passwordEditText.error = "must not be empty"
                if(TextUtils.isEmpty(password))
                    passwordEditText.error = "must not be empty"
            }
        }
    }

    private fun signIn(it: Any, email: String, password: String) {

    }







    fun validateData(email: String?, password: String): Boolean {
        //validate the data that are input by user
        if (!Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            emailEditText!!.error = "Email is invalid"
            return false
        }
        if (password.length < 6) {
            passwordEditText!!.error = "Password length is invalid"
            return false
        }
        return true
    }
}